operations/patch_util.go

```diff
@@ -368,100 +368,110 @@ func formatRev(rev string) string {
+		return ref
+	}
+	if commits != "" {
+		// if one commit, this returns just that commit, else the first commit in the range
+		return strings.Split(commits, "..")[0]
+	}
+	return "HEAD"
+}
+
 // loadGitData inspects the current git working directory and returns a patch and its summary.
 // The branch argument is used to determine where to generate the merge base from, and any extra
 // arguments supplied are passed directly in as additional args to git diff.
-func loadGitData(branch, ref, rev string, format bool, extraArgs ...string) (*localDiff, error) {
+func loadGitData(branch, ref, commits string, format bool, extraArgs ...string) (*localDiff, error) {
 	// branch@{upstream} refers to the branch that the branch specified by branchname is set to
 	// build on top of. This allows automatically detecting a branch based on the correct remote,
-	// if the user's repo is a fork, for example.
+	// if the user's repo is a fork, for example. This also works with a commit hash, if given.
+	// In the case a range is passed, we only need one commit to determine the base, so we use the last commit.
 	// For details see: https://git-scm.com/docs/gitrevisions
-	featureBranch := "HEAD"
-	if ref != "" {
-		featureBranch = ref
-	}
-	mergeBase, err := gitMergeBase(branch+"@{upstream}", featureBranch)
+
+	mergeBase, err := gitMergeBase(branch+"@{upstream}", ref, commits)
 	if err != nil {
 		return nil, errors.Wrap(err, "Error getting merge base")
 	}
 	statArgs := []string{"--stat"}
 	if len(extraArgs) > 0 {
 		statArgs = append(statArgs, extraArgs...)
 	}
-	stat, err := gitDiff(mergeBase, ref, rev, statArgs...)
+	stat, err := gitDiff(mergeBase, ref, commits, statArgs...)
 	if err != nil {
 		return nil, errors.Wrap(err, "Error getting diff summary")
 	}
-	log, err := gitLog(mergeBase, ref, rev)
+	log, err := gitLog(mergeBase, ref, commits)
 	if err != nil {
 		return nil, errors.Wrap(err, "git log")
 	}
 
 	var fullPatch string
 	if format {
-		fullPatch, err = gitFormatPatch(mergeBase, ref, rev)
+		fullPatch, err = gitFormatPatch(mergeBase, ref, commits)
 		if err != nil {
 			return nil, errors.Wrap(err, "Error getting formatted patch")
 		}
 	} else {
 		if !util.StringSliceContains(extraArgs, "--binary") {
 			extraArgs = append(extraArgs, "--binary")
 		}
-		fullPatch, err = gitDiff(mergeBase, ref, rev, extraArgs...)
+		fullPatch, err = gitDiff(mergeBase, ref, commits, extraArgs...)
 		if err != nil {
 			return nil, errors.Wrap(err, "Error getting patch")
 		}
 	}
 	return &localDiff{fullPatch, stat, log, mergeBase}, nil
 }
 
-// gitMergeBase runs "git merge-base <branch1> <branch2>" and returns the
-// resulting githash as string
-func gitMergeBase(branch1, branch2 string) (string, error) {
+// gitMergeBase runs "git merge-base <branch1> <branch2>" (where branch2 can optionally be a githash)
+// and returns the resulting githash as string
+func gitMergeBase(branch1, ref, commits string) (string, error) {
+	branch2 := getFeatureBranch(ref, commits)
```

Comment:
My thinking is:
For the commit queue we probably want the base to be the tip of upstream master.
For regular patches: for `...` we probably want the base to be the first commit, for `..` we probably want the merge base of the two commits, and for a single commit we probably want `commit~`

Reply:
You'll have to explain what you mean to me later. Are you saying that for a single commit As far as I can tell, using commit1 for any of `<commit1>..<commit2>` and `<commit1>...<commit2>` and `<commit1>` in gitFeatureBranch seems to work fine.

Reply:
I forgot that regular patches touched this. But regular patches don't have a commit flag? So we don't need to handle that case, it'll just return ref like it did before.